extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/DefaultTenantConfigResolver.java

```diff
@@ -377,4 +380,85 @@ private static ImmutablePathMatcher.ImmutablePathMatcherBuilder<String> addPath(
+        }
+
+        @Override
+        public String resolve(RoutingContext context) {
+            for (var tenantContext : tenantConfigContexts) {
+                final String token = OidcUtils.extractBearerToken(context, tenantContext.oidcConfig);
+                if (token != null && !OidcUtils.isOpaqueToken(token)) {
+                    final var tokenJson = OidcUtils.decodeJwtContent(token);
+                    if (tokenJson != null) {
+
+                        final String iss = tokenJson.getString(Claims.iss.name());
+                        if (tenantContext.getOidcMetadata().getIssuer().equals(iss)) {
+                            OidcUtils.storeExtractedBearerToken(context, token);
+
+                            final String tenantId = tenantContext.oidcConfig.tenantId.get();
+                            LOG.debugf("Resolved the '%s' OIDC tenant based on the matching issuer '%s'", tenantId, iss);
+                            return tenantId;
+                        }
+                    }
+                }
+            }
+            return null;
+        }
+
+        private static TenantResolver of(Map<String, TenantConfigContext> tenantConfigContexts) {
+            var contextsWithIssuer = new ArrayList<TenantConfigContext>();
+            for (TenantConfigContext context : tenantConfigContexts.values()) {
+                if (context.oidcConfig.tenantEnabled && !OidcUtils.isWebApp(context.oidcConfig)
+                        && context.getOidcMetadata() != null
+                        && context.getOidcMetadata().getIssuer() != null) {
+                    contextsWithIssuer.add(context);
+                }
+            }
+            if (contextsWithIssuer.isEmpty()) {
+                return null;
+            } else {
+                return new IssuerBasedTenantResolver(contextsWithIssuer.toArray(new TenantConfigContext[0]));
+            }
+        }
+
+        private static void addIssuerBasedTenantResolver(ArrayList<TenantResolver> resolvers,
+                Map<String, TenantConfigContext> staticTenantsConfig, TenantConfigContext defaultTenant) {
+            Map<String, TenantConfigContext> tenantConfigContexts = new HashMap<>(staticTenantsConfig);
+            tenantConfigContexts.put(OidcUtils.DEFAULT_TENANT_ID, defaultTenant);
+            var issuerTenantResolver = IssuerBasedTenantResolver.of(tenantConfigContexts);
+            if (issuerTenantResolver != null) {
+                resolvers.add(issuerTenantResolver);
+            } else {
+                LOG.warn("The 'quarkus.oidc.select-tenants-with-issuer' configuration property is set to true, "
+                        + "but no static tenant supports this feature. To use this feature, please configure at least "
+                        + "one static tenant with discovered issuer and set either 'service' or 'hybrid' application type");
+            }
+        }
+    }
+
+    private TenantResolver createAnnotationBasedTenantResolver() {
```

Comment:
This is a nice OO code, but IMHO creating a resolver instance for this check is not necessary and as I said, the same check is done during the dynamic resolution. I propose to keep an explicit set of steps, it is more verbose and not ideal but I can see easily what is going on, sorry about it

Reply:
@michalvavrik May be we can keep 4 others in the array if you'd like (Custom TenantResolver, configured-path-based-one, default path based one and the issuer) but please keep the annotations check separately and retain the restriction that a single custom tenantresolver can be used

Reply:
> This is a nice OO code, but IMHO creating a resolver instance for this check is not necessary and as I said, the same check is done during the dynamic resolution. I propose to keep an explicit set of steps, it is more verbose and not ideal but I can see easily what is going on, sorry about it

I am sorry @sberyozkin , but I need additional explanation here. Putting resolvers into form of resolvers allowed:

1. only run resolver when it is activate
2. reduce amount of IF conditions and run them only once - when the bean is created
3. IMHO simplification, code is simpler then IF-chain

So I don't understand pros, this is explicit in a sense that it runs only collected resolvers. Is that condition chain really more readable for you? I can respect that, even if I personally disagree.

> @michalvavrik May be we can keep 4 others in the array if you'd like (Custom TenantResolver, configured-path-based-one, default path based one and the issuer) but please keep the annotations check separately and retain the restriction that a single custom tenantresolver can be used

I am absolutely fine with doing what you are asking for, but truth be told, I don't understand it. Annotation based selection is no different to others. Maybe you could give me few hints. Also it is detectable during the build time, so I don't know why to keep it there when it is not detected.

Reply:
@michalvavrik The code related to the TENANT_ID check is also run if the dynamic resolver returns null, before the static resolution begins, so my point is that capturing it in the instance of `TenantResolver` is a bit too much, having a simple function called from both places won't be too bad IMHO.

As far as the code clarity is concerned, you may have a point, and I'm generally positive about cleaning up things, but I prefer it be done as part of dedicated PRs as otherwise the borders are getting blurred as to what exactly this PR focuses one and we have to start talking about things not related to the issue at hand.

In any case, I'm proposing a compromise, do not wrap the TENANT_ID check in the resolver, but indeed keep the rest of them in the array. Does it work for you ?

Thanks

Reply:
> As far as the code clarity is concerned, you may have a point, and I'm generally positive about cleaning up things, but I prefer it be done as part of dedicated PRs as otherwise the borders are getting blurred as to what exactly this PR focuses one and we have to start talking about things not related to the issue at hand.

Copy that.

> In any case, I'm proposing a compromise, do not wrap the TENANT_ID check in the resolver, but indeed keep the rest of them in the array. Does it work for you ?

Sounds good, I'm on it.